photon-lib/src/test/scala/com/linkedin/photon/ml/function/PriorDistributionTest.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.function
+
+import breeze.linalg.{DenseVector, diag}
+import org.testng.annotations.Test
+import org.testng.Assert.assertEquals
+import org.mockito.Mockito.mock
+
+import com.linkedin.photon.ml.model.{Coefficients => ModelCoefficients}
+import com.linkedin.photon.ml.normalization.NormalizationContext
+import com.linkedin.photon.ml.util.BroadcastWrapper
+
+/**
+ * Unit tests for [[PriorDistribution]], [[PriorDistributionDiff]], and [[PriorDistributionTwiceDiff]].
+ */
+class PriorDistributionTest {
+
+  import L2RegularizationTest._
+
+  private val DIMENSION = 4
+
+  /**
+   * Test that the prior distribution mixin traits can correctly modify the existing behaviour of an objective function.
+   */
+  @Test
+  def testAll(): Unit = {
+
+    val mockNormalization = mock(classOf[BroadcastWrapper[NormalizationContext]])
+
+    val coefficients = DenseVector.ones[Double](DIMENSION)
+    val priorMean = coefficients :* 2D
+    val multiplyVector = coefficients * 3D
+    val priorVar = coefficients :* 4D
+
+    val l1Weight = 10D
+    val l2Weight = 10D
+
+    val mockObjectiveFunction = new MockObjectiveFunction with PriorDistributionTwiceDiff {
+      override val priorCoefficients = ModelCoefficients(priorMean, Option(priorVar))
+      l1RegWeight = l1Weight
+      l2RegWeight = l2Weight
+    }
+
+    // Assume that coefficients = 1-vector, prior mean = 2-vector, multiply = 3-vector, prior variance = 4-vector for all expected values below
+    val expectedValue = MockObjectiveFunction.VALUE + 0.25 * l2Weight * DIMENSION / 2 + 0.5 * l1Weight * DIMENSION
```

Comment:
You might want to add some comments explaining the mathematics, or do more computations with the raw variables. There's some skipped steps/simplified values in these calculations which confused me for a bit, even though I'm looking at the design document.

Reply:
Fixed. Add more comments in `PriorDistribution.scala` and `PriorDistributionTest.scala`